core/drivers/clk/clk.c

```diff
@@ -0,0 +1,615 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stddef.h>
+
+#include <drivers/scmi.h>
+
+#define SCMI_MAX_CLK_NAME_LEN	16
+
+static SLIST_HEAD(, clk) clk_list = SLIST_HEAD_INITIALIZER(clk_list);
+
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, uint8_t parent_count)
+{
+	struct clk *clk;
+	unsigned int parent;
+	unsigned int alloc_size = sizeof(struct clk) +
+				  parent_count * sizeof(struct clk *);
+
+	clk = malloc(alloc_size);
+	if (!clk)
+		panic();
+
+	memset(clk, 0, alloc_size);
```

Comment:
prefer `calloc(1, alloc_size)` and remove this `memset()`.

Reply:
Done